server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -99,6 +103,21 @@ protected boolean localExecute(Request request) {
 
     protected abstract ClusterBlockException checkBlock(Request request, ClusterState state);
 
+    private ClusterBlockException checkBlockIfStateRecovered(Request request, ClusterState state) {
+        try {
+            return checkBlock(request, state);
+        } catch (IndexNotFoundException e) {
+            if (state.blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+                // no index metadata is exposed yet, but checkBlock depends on an index, so keep trying until the cluster forms
+                assert GatewayService.STATE_NOT_RECOVERED_BLOCK.contains(ClusterBlockLevel.METADATA_READ);
+                assert state.blocks().global().stream().allMatch(ClusterBlock::retryable);
```

Comment:
Sorry I gave you a duff assertion here, should be this:

```suggestion
                assert state.blocks().global(ClusterBlockLevel.METADATA_READ).stream().allMatch(ClusterBlock::retryable);
```

(or else we could theoretically have something like `Metadata.CLUSTER_READ_ONLY_BLOCK` in there which isn't retryable but also doesn't block metadata reads).